### DISCUSSION TITLE
Stronger type-safety for queries

### DISCUSSION THREAD

**Discussion by ghost** [mikro-orm/mikro-orm#2231] [Ideas]
Hey,
So far, I've been enjoying MikroORM. It's amazing and repositories make my code very organized. I was just wondering if there's a way to get more accurate type-safety with queries. I'll share a code snippet.
```ts
@Entity()
export class Infraction {
  @Property()
  name: string;

  @Property({ type: Date })
  issuedAt = new Date();

  @Property({ optional: true })
  duration: number;
}
```
Then in our repository, we have a method to extract expired records.
```ts
@Repository(Infraction)
export class InfractionRepository extends EntityRepository<Infraction> {
  findDiscardable() {
    const qb = this.createQueryBuilder();
    return qb
      .select('*')
      .where({
        penalty: { $in: EXPIRABLE_INFRACTIONS },
        duration: { $lte: qb.raw('now() - e.issued_at') }
      })
      .getResult();
  }
}
```
If I run `infractionRepository.findDiscardable()`, the `Promise` it returns is just an array of generic infractions (`Promise<Infraction[]>`). Are there any updates for MikroORM to have stronger and more accurate type-safety soon? It might be complex to implement, but ideally the results should reflect the conditions of the query itself. In this case, `duration` should be a required field and `penalty` should be a union type of the elements of the `EXPIRABLE_INFRACTIONS` array. Apologies if this is too much to ask for, but I think it could make the ORM a lot more type-safe and reliable.

**Comment by B4nan** [comment]
That sounds quite extreme to implement, as the conditions can be nested, and don't let me started about understanding operators (not talking about the fact that you are mixing it with raw queries). Too many moving parts. 

We do have `Loaded` type for EM methods, that respects the `populate` parameter (marks relations as loaded in case of `Reference` or `Collection` properties), I don't really plan to make it smarter right now, it already is quite complex and performance heavy. 

If you are into advanced TS mapping, feel free to try to improve it. Personally I would say it is not worth it, it sounds impossible to cover all possibilities (and it would be still inaccurate). Better to define your own type where you will be 100% sure what it does and it will not add any performance overhead.

**Reply by ghost** [reply]
I understand. I myself am not proficient enough with TypeScript to work on advanced type mappings where you conditionally alter properties of classes so for now, I think I might just have to type-cast it to my custom type. I'll still look in to opening a PR if I ever find out how to do it though. Thanks for the help!
